adapter: Limit amount of data we buffer for COPY FROM

ParkMyCar opened the pull request:
This PR limits the amount of data we buffer for the `COPY FROM` command. Currently when using `COPY FROM` we buffer all of the data into memory, which could cause environmentd to OOM if a user tries to copy more than 4GB, which is how much memory we allocate to the `environmentd` pod.

Note: the max_results_size param is configurable via [LaunchDarkly](https://app.launchdarkly.com/default/production/features/max_result_size/targeting)

### Motivation

There was a [discussion in #feedback in Slack](https://materializeinc.slack.com/archives/CU7ELJ6E9/p1694557022323679) that talked about copying large amounts of data, which made me think about this possibility.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  <!-- Reference the design in the description. -->
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - Sets a max amount of data that the user is able send with `COPY FROM`.

benesch commented:
This is such a good idea! I think we should introduce a separate parameter for this, though. Using the same default value as `max_result_size` makes sense, I think, but would be nice not tie these two limits together.

ParkMyCar commented:
> This is such a good idea! I think we should introduce a separate parameter for this, though. Using the same default value as `max_result_size` makes sense, I think, but would be nice not tie these two limits together.

Makes sense! I updated the param to be `max_copy_from_size` and updated the docs to call out the limit

jkosh44 commented on src/sql/src/session/vars.rs:
```diff
@@ -1725,6 +1733,7 @@ impl SessionVars {
             )
             .with_var(&MAX_QUERY_RESULT_SIZE)
             .with_var(&MAX_IDENTIFIER_LENGTH)
+            .with_var(&MAX_COPY_FROM_SIZE)
```
Discussed in person, but writing here for the record, I think we should probably make this a system var like `MAX_RESULT_SIZE`. That way a single session can't decide to OOM environmentd.

ParkMyCar replied:
Good call, updated to make `max_copy_from_size` a SystemVar

jkosh44 reviewed (approved):
LGTM, can you add a test for this.
